subprojects/core/src/main/java/org/gradle/api/internal/attributes/IncubatingAttributesChecker.java

```diff
@@ -0,0 +1,84 @@
+
+import org.gradle.api.Incubating;
+import org.gradle.api.attributes.Attribute;
+import org.gradle.api.attributes.AttributeContainer;
+import org.gradle.api.attributes.Category;
+import org.gradle.internal.Pair;
+
+import javax.annotation.Nullable;
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Static util class which provides utilities to check if an {@link Attribute} is {@link org.gradle.api.Incubating}.
+ *
+ * An attribute is incubating iff it is a type annotated with {@link org.gradle.api.Incubating}, or if the value is equal to the value of a constant defined
+ * in the corresponding attribute interface which is annotated with {@link org.gradle.api.Incubating}.
+ *
+ * For example, any attribute named {@link org.gradle.api.attributes.TestSuiteType} is currently incubating, as is a {@link Category#CATEGORY_ATTRIBUTE} with a value
+ * equal to {@link Category#VERIFICATION}.
+ *
+ * @since 7.5
+ */
+public abstract class IncubatingAttributesChecker {
+    public static <T> boolean isIncubating(Attribute<T> attribute) {
+        return isIncubatingAttributeValue(attribute.getType(), null);
+    }
+
+    public static <T> boolean isIncubating(Attribute<T> key, @Nullable T value) {
+        if (isIncubatingAttributeInterface(key)) {
+            return true;
+        } else {
+            return isIncubatingAttributeValue(key.getType(), value);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public static boolean isAnyIncubating(AttributeContainer attributes) {
+        return attributes.keySet().stream().map(k -> Pair.of((Attribute<Object>) k, (Object) attributes.getAttribute(k))).anyMatch(IncubatingAttributesChecker::isIncubating);
+    }
+
+    private static <T> boolean isIncubating(Pair<Attribute<T>, T> attributePair) {
+        return isIncubating(Objects.requireNonNull(attributePair.getLeft()), attributePair.getRight());
+    }
+
+    private static <T> boolean isIncubatingAttributeInterface(Attribute<T> key) {
+        return Arrays.stream(key.getType().getAnnotations()).anyMatch(Incubating.class::isInstance);
+    }
+
+    private static <T> boolean isIncubatingAttributeValue(Class<T> type, @Nullable T value) {
+        return getIncubatingFields(type).stream().map(IncubatingAttributesChecker::getFieldValue).anyMatch(v -> Objects.equals(v, value != null ? value.toString() : null));
+    }
+
+    private static Object getFieldValue(Field f) {
+        try {
+            return f.get(null);
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException("Error reading field: " + f.getName() + " on type: " + f.getType().getName(), e);
+        }
+    }
+
+    private static List<Field> getIncubatingFields(Class<?> type) {
+        return Arrays.stream(type.getDeclaredFields()).filter(f -> Arrays.stream(f.getAnnotations()).anyMatch(Incubating.class::isInstance)).collect(Collectors.toList());
```

Comment:
Not sure if it'd be worth pulling out the `Arrays.stream(f.getAnnotations()).anyMatch(Incubating.class::isInstance)` to a separate method, since you also do it in `isIncubatingAttributeInterface` -- I believe you could pull out a common method using `AnnotatedElement`.

Also, this can use `f.isAnnotationPresent(Incubating.class)` rather than iterating over everything. You could do this in the extracted method too.

Reply:
Good ideas, worth consolidating this I think.